File: 2020/day21/python/day21.py
#!/usr/bin/env python3
from os.path import dirname
from os.path import realpath
from os.path import join
import time
import pytest
import sys

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))


def transform_input(input_):
    # custom transform for the day

    ingredients_list, allergens_list = [], []
    for inp in input_.splitlines():

        # Split up the parts
        ingredients, allergens = inp.split("(contains ")

        # Remove white spaces around ingredients and the allergens
        ingredients = ingredients.split()
        allergens = allergens[:-1].replace(" ", "").split(",")

        # Append them to the list
        ingredients_list.append(ingredients)
        allergens_list.append(allergens)

    return [ingredients_list, allergens_list]


def read_input(file_name="../data/input.txt"):
    dir_path = dirname(realpath(__file__))
    with open(join(dir_path, file_name), "r") as f:
        input_ = transform_input(f.read())

    return input_


def match_allergens(input_):

    ingredients_list, allergens_list = input_

    possible_allergen = dict()
    all_ingredients = set()

    # Go through the list
    for allergens, ingredients in zip(allergens_list, ingredients_list):

        # Record all ingredients we've seen
        all_ingredients |= set(ingredients)

        # Go through each allergen
        for allergen in allergens:
            # If we haven't seen this allergen before, we know it can only be
            # produced by this ingredients
            if allergen not in possible_allergen:
                possible_allergen[allergen] = set(ingredients)
            # Otherwise, only keep ingredients we've seen before
            else:
                possible_allergen[allergen] = possible_allergen[
                    allergen
                ].intersection(ingredients)
    # Returns a set of all ingredients, and which may cause allergies
    return all_ingredients, possible_allergen


def solve_part1(input_):

    all_ingredients, possible_allergen = match_allergens(input_)

    # Remove all ingredients that can cause an allergy
    for ingredients in possible_allergen.values():
        all_ingredients -= ingredients

    # Go through the ingredients list and see how often the ingredients appear
    return sum(
        [
            1
            for ingredients in input_[0]
            for ingredient in ingredients
            if ingredient in all_ingredients
        ]
    )


def solve_part2(input_):

    all_ingredients, possible_allergen = match_allergens(input_)

    match = dict()
    # Keep going until we have matched as many ingredients as allergens
    while len(match) < len(possible_allergen):
        for allergy in possible_allergen:
            # If the allergy only has one ingredient, we match the ingredient
            if len(possible_allergen[allergy]) == 1:
                match[allergy] = possible_allergen[allergy].pop()
                # Remove the ingredient from the other possible reactions
                for allergy_ in possible_allergen:
                    if match[allergy] in possible_allergen[allergy_]:
                        possible_allergen[allergy_].remove(match[allergy])

    return "".join(
        [match[key] + "," for key in sorted(list(possible_allergen.keys()))]
    )[:-1]


def main():
    input_ = read_input()

    t0 = time.time()
    part1 = solve_part1(input_)
    time_part1 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part1} (time taken {time_part1}[ms])")

    t0 = time.time()
    part2 = solve_part2(input_)
    time_part2 = round((time.time() - t0) * 1e3)
    print(f"Solution to part two: {part2} (time taken {time_part2}[ms])")


if __name__ == "__main__":
    main()


@pytest.mark.parametrize("input1, output1", [("../data/test_input0.txt", 5)])
def test1(input1, output1):
    assert solve_part1(read_input(input1)) == output1


@pytest.mark.parametrize(
    "input2, output2", [("../data/test_input0.txt", "mxmxvkd,sqjhc,fvjkl")]
)
def test2(input2, output2):
    assert solve_part2(read_input(input2)) == output2

File: 2020/day21/cpp/src/day21.cpp
#include "day21.hpp"

#include <boost/algorithm/string.hpp>

#include "utils.hpp"

using day21::input_type;
using day21::map_type;
using day21::match_type;
using day21::output_type1;
using day21::output_type2;
using day21::set_type;

input_type day21::transform_input(std::vector<std::string> &input) {
  input_type output;

  for (auto &inp : input) {
    // Split up the parts
    std::vector<std::string> split;
    utils::split_string(inp, "(contains ", split);

    // Remove white spaces around ingredients and the allergens
    std::string left(split[0]), right(split[1]);
    boost::algorithm::trim(left);
    boost::algorithm::trim(right);

    // Filter ingredients
    std::vector<std::string> ingredients;
    utils::split_string(left, " ", ingredients);

    // Filter allergens (and remove last parenthesis)
    right.erase(std::remove(right.begin(), right.end(), ' '), right.end());
    std::vector<std::string> allergens;
    utils::split_string(right.substr(0, right.length() - 1), ",", allergens);

    // Append them to the output
    output.push_back(std::make_pair(ingredients, allergens));
  }

  return output;
}

input_type day21::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

match_type day21::match_allergens(const input_type &input) {
  map_type possible_allergen;
  set_type all_ingredients;

  // Go through the list
  for (auto &[ingredients, allergens] : input) {
    // Record all ingredients we've seen
    set_type ingredients_set(ingredients.begin(), ingredients.end());
    all_ingredients.insert(ingredients_set.begin(), ingredients_set.end());

    // Go through each allergen
    for (auto allergen : allergens) {
      // If we haven't seen this allergen before, we know it can only be
      // produced by these ingredients
      if (!possible_allergen.count(allergen)) {
        possible_allergen[allergen] = ingredients_set;
        // Otherwise, only keep ingredients we've seen before
      } else {
        std::vector<std::string> o;
        std::set_intersection(possible_allergen[allergen].begin(),
                              possible_allergen[allergen].end(),
                              ingredients_set.begin(), ingredients_set.end(),
                              std::back_insert_iterator(o));
        possible_allergen[allergen] = set_type({o.begin(), o.end()});
      }
    }
  }
  // Returns a set of all ingredients, and which may cause allergies
  return match_type({all_ingredients, possible_allergen});
}

// Implement possible_allergenrt one solution
output_type1 day21::solve_part1(const input_type &input_) {
  auto [all_ing, poss_all] = match_allergens(input_);

  // Remove all ingredients that can cause an allergy
  for (auto &k : poss_all) {
    auto &ingreds = k.second;
    set_type diff;
    std::set_difference(all_ing.begin(), all_ing.end(), ingreds.begin(),
                        ingreds.end(), std::inserter(diff, diff.begin()));
    all_ing = diff;
  }

  // Go through the ingredients list and see how often the ingredients appear
  int s(0);
  for (auto &k : input_) {
    for (auto &ing : k.first) {
      if (all_ing.count(ing)) {
        s++;
      }
    }
  }
  return s;
}

// Implement part two solution
output_type2 day21::solve_part2(const input_type &input_) {
  auto possible_allergen = match_allergens(input_).second;

  // Keep going until we have matched as many ingredients as allergens
  map_type match;
  while (match.size() < possible_allergen.size()) {
    for (auto &kv : possible_allergen) {
      auto &allergy = kv.first;
      // If the allergy only has one ingredient, we match the ingredient
      if (possible_allergen[allergy].size() == 1) {
        std::string v = *possible_allergen[allergy].begin();
        match[allergy] = set_type({v});
        possible_allergen[allergy].erase(v);
        // If the allergy only has one ingredient, we match the ingredient
        for (auto &kv_ : possible_allergen) {
          auto &allergy_ = kv_.first;
          if (possible_allergen[allergy_].count(v)) {
            possible_allergen[allergy_].erase(v);
          }
        }
      }
    }
  }

  // Sort the allergens, and return the ingredients
  std::string msg;
  for (auto &k : match) {
    msg += *match[k.first].begin() + ",";
  }
  return msg.substr(0, msg.length() - 1);
}
